### PR TITLE
[SPARK-35683][PYTHON] Fix Index.difference to avoid collect 'other' to driver side

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR fix the wrong behavior of `Index.difference` in pandas APIs on Spark, based on the comment https://github.com/databricks/koalas/pull/1325#discussion_r647889901 and https://github.com/databricks/koalas/pull/1325#discussion_r647890007
- it couldn't handle the case properly when `self` is `Index` or `MultiIndex` and `other` is `MultiIndex` or `Index`.
```python
>>> midx1 = ps.MultiIndex.from_tuples([('a', 'x', 1), ('b', 'z', 2), ('k', 'z', 3)])
>>> idx1 = ps.Index([1, 2, 3])
>>> midx1 = ps.MultiIndex.from_tuples([('a', 'x', 1), ('b', 'z', 2), ('k', 'z', 3)])
>>> midx1.difference(idx1)
pyspark.pandas.exceptions.PandasNotImplementedError: The method `pd.Index.__iter__()` is not implemented. If you want to collect your data as an NumPy array, use 'to_numpy()' instead.
```
- it's collecting the all data into the driver side when the other is list-like objects, especially when the `other` is distributed object such as Series which is very dangerous.

And added the related test cases.

### Why are the changes needed?

To correct the incompatible behavior with pandas, and to prevent the case which potentially cause the OOM easily.

```python
>>> midx1 = ps.MultiIndex.from_tuples([('a', 'x', 1), ('b', 'z', 2), ('k', 'z', 3)])
>>> idx1 = ps.Index([1, 2, 3])
>>> midx1 = ps.MultiIndex.from_tuples([('a', 'x', 1), ('b', 'z', 2), ('k', 'z', 3)])
>>> midx1.difference(idx1)
MultiIndex([('a', 'x', 1),
            ('b', 'z', 2),
            ('k', 'z', 3)],
           )
```

And now it only using the for loop when the `other` is only the case `list`, `set` or `dict`.

### Does this PR introduce _any_ user-facing change?

Yes, the previous bug is fixed as described in the above code examples.

### How was this patch tested?

Manually tested with linter and unittest in local, and it might be passed on CI.

### CODE REVIEW

File: python/pyspark/pandas/indexes/base.py [hunk 1]
```diff
@@ -1992,6 +1992,16 @@ def difference(self, other: "Index", sort: Optional[bool] = None) -> "Index":
...
+            if isinstance(self, MultiIndex):
+                # In case `self` is MultiIndex and `other` is Index,
+                # return MultiIndex without its names.
+                return self.rename([None] * len(self))
+            elif isinstance(self, Index):
+                # In case `self` is Index and `other` is MultiIndex,
+                # return Index without its name.
+                return self.rename(None)
+
         if not is_list_like(other):
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/indexes/base.py:2005]: Can we don't depend on the pandas comparison? `is_list_like` in pandas is not designed to handle pandas-on-Spark instances, and I don't know if it's safe to use it against pandas-on-Spark instances.

File: python/pyspark/pandas/indexes/base.py [hunk 2]
```diff
@@ -1992,6 +1992,16 @@ def difference(self, other: "Index", sort: Optional[bool] = None) -> "Index":
         """
         from pyspark.pandas.indexes.multi import MultiIndex
 
+        if isinstance(other, Index) and not isinstance(self, type(other)):
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/indexes/base.py:1995]: can you extract this condition and define it in a variable? It's difficult to follow the if-elses here.

File: python/pyspark/pandas/indexes/base.py [hunk 3]
```diff
@@ -2000,11 +2010,16 @@ def difference(self, other: "Index", sort: Optional[bool] = None) -> "Index":
                     sort
                 )
             )
-        # Handling MultiIndex
+        # Handling MultiIndex when `other` is list-like objects.
         if isinstance(self, MultiIndex) and not isinstance(other, MultiIndex):
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/indexes/base.py:2014]: can you extract this condition and define it in a variable? It's difficult to follow the if-elses here.

**Comment by HyukjinKwon** [comment]
BTW, let's make a PR to Koalas side too since this is a bug fix.

**Comment by itholic** [comment]
Thanks for the review, @HyukjinKwon 

Let me create a PR when this is ready to merge.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**